File: backend/app/schemas/avisos.py
"""Schemas Pydantic para el modulo de Avisos (C-15).

Todos los schemas usan ``extra='forbid'`` (REGLAS DURAS #5).
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Aviso schemas ──────────────────────────────────────────────────────


class AvisoCreate(BaseModel):
    """Datos para crear un nuevo aviso institucional."""

    model_config = ConfigDict(extra="forbid")

    alcance: str
    materia_id: UUID | None = None
    cohorte_id: UUID | None = None
    rol_destino: str | None = None
    severidad: str
    titulo: str = Field(..., min_length=1, max_length=200)
    cuerpo: str = Field(..., min_length=1)
    inicio_en: datetime
    fin_en: datetime
    orden: int = Field(default=0, ge=0)
    requiere_ack: bool = False


class AvisoUpdate(BaseModel):
    """Datos para actualizar un aviso existente."""

    model_config = ConfigDict(extra="forbid")

    alcance: str | None = None
    materia_id: UUID | None = None
    cohorte_id: UUID | None = None
    rol_destino: str | None = None
    severidad: str | None = None
    titulo: str | None = Field(None, min_length=1, max_length=200)
    cuerpo: str | None = Field(None, min_length=1)
    inicio_en: datetime | None = None
    fin_en: datetime | None = None
    orden: int | None = Field(None, ge=0)
    activo: bool | None = None
    requiere_ack: bool | None = None


class AvisoResponse(BaseModel):
    """Respuesta completa de un aviso con metricas de tracking."""

    model_config = ConfigDict(extra="forbid")

    id: UUID
    alcance: str
    materia_id: UUID | None = None
    cohorte_id: UUID | None = None
    rol_destino: str | None = None
    severidad: str
    titulo: str
    cuerpo: str
    inicio_en: datetime
    fin_en: datetime
    orden: int
    activo: bool
    requiere_ack: bool
    created_at: str | None = None
    updated_at: str | None = None
    total_ack: int = 0
    total_usuarios_alcance: int = 0
    porcentaje_ack: float = 0.0


# ── Acknowledgment schemas ─────────────────────────────────────────────


class AcknowledgmentCreate(BaseModel):
    """Datos para confirmar lectura de un aviso."""

    model_config = ConfigDict(extra="forbid")

    aviso_id: UUID


class AcknowledgmentResponse(BaseModel):
    """Respuesta de un acknowledgment."""

    model_config = ConfigDict(extra="forbid")

    id: UUID
    aviso_id: UUID
    usuario_id: UUID
    confirmado_at: datetime
    usuario_nombre: str | None = None
    usuario_email: str | None = None


# ── List/Tracking schemas ──────────────────────────────────────────────


class AvisoTimelineItem(BaseModel):
    """Item de la timeline de avisos para un usuario."""

    model_config = ConfigDict(extra="forbid")

    id: UUID
    alcance: str
    severidad: str
    titulo: str
    cuerpo: str
    inicio_en: datetime
    fin_en: datetime
    orden: int
    requiere_ack: bool
    acknowledged: bool = False
    created_at: str | None = None


class AvisoTimelineResponse(BaseModel):
    """Timeline de avisos activos para el usuario autenticado."""

    model_config = ConfigDict(extra="forbid")

    items: list[AvisoTimelineItem]
    total: int


class TrackingAckItem(BaseModel):
    """Item del tracking de acknowledgments."""

    model_config = ConfigDict(extra="forbid")

    usuario_id: UUID
    usuario_nombre: str | None = None
    confirmado_at: datetime | None = None


class TrackingAvisoResponse(BaseModel):
    """Tracking de acknowledgments de un aviso."""

    model_config = ConfigDict(extra="forbid")

    total_usuarios: int = 0
    total_ack: int = 0
    porcentaje: float = 0.0
    acknowledgments: list[TrackingAckItem] = []

File: frontend/src/features/avisos/types/avisos.ts
import { z } from "zod";

export const AvisoCreateSchema = z.object({
  alcance: z.enum(["Global", "PorMateria", "PorCohorte", "PorRol"]),
  materia_id: z.string().uuid().optional(),
  cohorte_id: z.string().uuid().optional(),
  rol_destino: z.string().optional(),
  severidad: z.enum(["Info", "Advertencia", "Crítico"]),
  titulo: z.string().min(1).max(200),
  cuerpo: z.string().min(1),
  inicio_en: z.string(),
  fin_en: z.string(),
  orden: z.number().int().default(0),
  requiere_ack: z.boolean().default(false),
}).strict();

export type AvisoCreate = z.infer<typeof AvisoCreateSchema>;

export const AvisoUpdateSchema = z.object({
  alcance: z.enum(["Global", "PorMateria", "PorCohorte", "PorRol"]).optional(),
  materia_id: z.string().uuid().optional(),
  cohorte_id: z.string().uuid().optional(),
  rol_destino: z.string().optional(),
  severidad: z.enum(["Info", "Advertencia", "Crítico"]).optional(),
  titulo: z.string().min(1).max(200).optional(),
  cuerpo: z.string().min(1).optional(),
  inicio_en: z.string().optional(),
  fin_en: z.string().optional(),
  orden: z.number().int().optional(),
  requiere_ack: z.boolean().optional(),
  activo: z.boolean().optional(),
}).strict();

export type AvisoUpdate = z.infer<typeof AvisoUpdateSchema>;

export const AvisoResponseSchema = z.object({
  id: z.string().uuid(),
  alcance: z.string(),
  materia_id: z.string().uuid().nullable().optional(),
  cohorte_id: z.string().uuid().nullable().optional(),
  rol_destino: z.string().nullable().optional(),
  severidad: z.string(),
  titulo: z.string(),
  cuerpo: z.string(),
  inicio_en: z.string(),
  fin_en: z.string(),
  orden: z.number().int(),
  activo: z.boolean(),
  requiere_ack: z.boolean(),
  created_at: z.string().nullable().optional(),
  updated_at: z.string().nullable().optional(),
  total_ack: z.number().int(),
  total_usuarios_alcance: z.number().int(),
  porcentaje_ack: z.number(),
});

export type AvisoResponse = z.infer<typeof AvisoResponseSchema>;

export const AvisoTimelineItemSchema = z.object({
  id: z.string().uuid(),
  alcance: z.string(),
  severidad: z.string(),
  titulo: z.string(),
  cuerpo: z.string(),
  inicio_en: z.string(),
  fin_en: z.string(),
  orden: z.number().int(),
  requiere_ack: z.boolean(),
  acknowledged: z.boolean(),
  created_at: z.string().nullable().optional(),
});

export type AvisoTimelineItem = z.infer<typeof AvisoTimelineItemSchema>;

export const TrackingAvisoResponseSchema = z.object({
  total_usuarios: z.number().int(),
  total_ack: z.number().int(),
  porcentaje: z.number(),
  acknowledgments: z.array(
    z.object({
      usuario_id: z.string().uuid(),
      usuario_nombre: z.string().nullable().optional(),
      confirmado_at: z.string().nullable().optional(),
    }),
  ),
});

export type TrackingAvisoResponse = z.infer<typeof TrackingAvisoResponseSchema>;

export interface AvisoFilters {
  materia_id?: string;
  cohorte_id?: string;
  alcance?: string;
  severidad?: string;
  activo?: boolean;
}
